instrumentation/opentelemetry-instrumentation-tornado/src/opentelemetry/instrumentation/tornado/client.py

```diff
@@ -109,13 +109,14 @@ def _finish_tracing_callback(
     description = None
     exc = future.exception()
 
-    response = future.result()
-
     if span.is_recording() and exc:
         if isinstance(exc, HTTPError):
             status_code = exc.code
         description = f"{type(exc).__name__}: {exc}"
-    else:
+
+    response = None
+    if not exc:
```

Comment:
I'm not too familiar with tornado but what is the relationship between future.exception() and future.result()? Previously, the behavior was that `result()` was called regardless of if there is an exception. Was that actually buggy behavior?

Reply:
We have unfortunately not a consistent way of handling requests but we still require instrumentations to capture information if an exception is thrown. For example, the `requests` instrumentation [captures information about the exception](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L277) and [reraises](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L303) it to be [recorded](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L993) on span exit. Similarly, other instrumentations like aiohttpclient [records the exception manually](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/3669b477bdf2c8631ab223599baa4b34ee18cf0d/instrumentation/opentelemetry-instrumentation-aiohttp-client/src/opentelemetry/instrumentation/aiohttp_client/__init__.py#L242). With this change, in the presence of an exception, it looks like recording will skipped?

Reply:
> I'm not too familiar with tornado but what is the relationship between future.exception() and future.result()

The problem is that if the future has an exception result, then calling future.result() will throw that exception. The previous code did not cover this case, future.result() was called without an try-catch, so when it raised the exception it exited the method and bubbled up to the async runtime (since it's not caught higher in the stack either).

I think this fix is an improvement on how it was, by making sure we don't raise an exception that goes uncaught here. But we may be able to get the response out of the exception, I'll look into how this works with tornado.

Reply:
@lzchen latest changes fulfil this